kubetest2/kubetest2-gke/deployer/deployer.go

```diff
@@ -0,0 +1,516 @@
+	if err := d.getInstanceGroups(); err != nil {
+		return "", err
+	}
+	// We want to ensure that there's an e2e-ports-* firewall rule
+	// that maps to the cluster nodes, but the target tag for the
+	// nodes can be slow to get. Use the hash from the lexically first
+	// node pool instead.
+	return "e2e-ports-" + d.instanceGroups[0].uniq, nil
+}
+
+// This function ensures that all firewall-rules are deleted from specific network.
+// We also want to keep in logs that there were some resources leaking.
+func (d *deployer) cleanupNetworkFirewalls() (int, error) {
+	fws, err := exec.Output(exec.Command("gcloud", "compute", "firewall-rules", "list",
+		"--format=value(name)",
+		"--project="+d.project,
+		"--filter=network:"+d.network))
+	if err != nil {
+		return 0, fmt.Errorf("firewall rules list failed: %s", execError(err))
+	}
+	if len(fws) > 0 {
+		fwList := strings.Split(strings.TrimSpace(string(fws)), "\n")
+		log.Printf("Network %s has %v undeleted firewall rules %v", d.network, len(fwList), fwList)
+		commandArgs := []string{"compute", "firewall-rules", "delete", "-q"}
+		commandArgs = append(commandArgs, fwList...)
+		commandArgs = append(commandArgs, "--project="+d.project)
+		errFirewall := runWithOutput(exec.Command("gcloud", commandArgs...))
+		if errFirewall != nil {
+			return 0, fmt.Errorf("error deleting firewall: %v", errFirewall)
+		}
+		return len(fwList), nil
+	}
+	return 0, nil
+}
+
+func (d *deployer) Down() error {
+	if err := d.verifyFlags(); err != nil {
+		return err
+	}
+	if err := d.prepareGcpIfNeeded(); err != nil {
+		return err
+	}
+	firewall, err := d.getClusterFirewall()
+	if err != nil {
+		// This is expected if the cluster doesn't exist.
+		return nil
+	}
+	d.instanceGroups = nil
+
+	loc, err := d.location()
+	if err != nil {
+		return err
+	}
+
+	// We best-effort try all of these and report errors as appropriate.
+	errCluster := runWithOutput(exec.Command(
+		"gcloud", d.containerArgs("clusters", "delete", "-q", d.cluster,
+			"--project="+d.project,
+			loc)...))
+
+	// don't delete default network
+	if d.network == "default" {
+		if errCluster != nil {
+			log.Printf("Error deleting cluster using default network, allow the error for now %s", errCluster)
+		}
+		return nil
+	}
+
+	var errFirewall error
+	if runWithNoOutput(exec.Command("gcloud", "compute", "firewall-rules", "describe", firewall,
+		"--project="+d.project,
+		"--format=value(name)")) == nil {
+		log.Printf("Found rules for firewall '%s', deleting them", firewall)
+		errFirewall = exec.Command("gcloud", "compute", "firewall-rules", "delete", "-q", firewall,
+			"--project="+d.project).Run()
+	} else {
+		log.Printf("Found no rules for firewall '%s', assuming resources are clean", firewall)
+	}
+	numLeakedFWRules, errCleanFirewalls := d.cleanupNetworkFirewalls()
+	var errSubnet error
+	errNetwork := runWithOutput(exec.Command("gcloud", "compute", "networks", "delete", "-q", d.network,
+		"--project="+d.project))
+	if errCluster != nil {
```

Comment:
we should consider using a multi-error type to collect up and return all of these in the future

Reply:
https://github.com/kubernetes/apimachinery/blob/b74e4c97951f106696db86d7593effcb1463ca30/pkg/util/errors/errors.go#L31

Reply:
consider using a type like that (k8s.io/apimachinery/pkg/util/errors.Aggregate)

Reply:
(can be done in a follow-up)